### ISSUE TITLE
NPC Crafting

### ISSUE THREAD

**Issue by eltank** [CleverRaven/Cataclysm-DDA#51628]
Writing down ideas for a feature I'm working on.

# Is your feature request related to a problem? Please describe.
A) NPCs can't raise their practical skill levels because they can't craft practice recipes. In fact they can't craft anything at all on their own.
B) If you have enough tools and materials and you have N skilled helpers you should be able to craft N+1 recipes at the same time.

Problems to overcome:
1) Most recipes require choosing ingredients at the start, which is something the player needs to decide.
2) When crafting fails, the player needs to decide whether it's worth spending ingredients to continue.
3) When multiple characters are crafting at the same time, they shouldn't use the same tools or the same workbench
4) Tools running out of charge shouldn't require micro-management
5) Practice recipes should be repeated until level-up.
6) Make sure NPCs become weary (and suffer the consequences).
7) Handling liquid result/byproducts needs player input.
8) Crafting is not simulated outside the reality bubble.

# Describe the solution you'd like
- Define a Crafting Zone.
- Ask NPCs to work on in-progress crafts in the Crafting Zone.

1) There is an easy way to handle it. The player starts crafting, then pauses the activity, leaving the in-progress craft on the workbench, which can then be finished by an NPC.

2) This one is easy. The NPC should simply stop crafting on failure. The player can interact with the in-progress craft to continue, then pause the activity and let the NPC finish.

3) When an NPC starts crafting, reserve the tools that they need using item vars. I believe this is already done for other activities like construction, vehicle repair and vehicle disassembly. Additionally, reserve the crafting area (workbench). NPCs who can't find tools or an available crafting area in the zone will cancel the crafting activity. The player should always be able to take tools that they need from NPCs, which will also cancel their activity.

4) NPCs (and the avatar) should automatically reload battery-powered tools if they run out while crafting.

5) This should be easy to do, since repeating a craft doesn't require choosing ingredients again. When crafting practice recipes, the default behavior should be to either repeat until level-up or repeat until skill can't be gained any more.

6) Weariness is handled in character.cpp so hopefully it just works for NPCs, needs testing.

7) The easiest solution is to stop crafting just before completion (or if possible leave the in-progress craft at 100%) if the products can't simply be placed on the workbench. Then the player can come along and finish it. A more involved solution is to fill available containers.

8) Ideally, perform catch-up simulation when a crafting NPC re-enters the reality bubble. That's not exactly trivial, so for an initial implementation just pause the activity while outside the reality bubble.

# Describe alternatives you've considered

1) Make it possible to "prepare" a craft without actually starting it.

1,2) Can also be handled by taking control of an NPC, starting the craft then switching to another NPC. But I want to have a solution that works outside of party-based gameplay.

# Additional context

Hmm, the solution for 5 gave me an idea regarding the book reading activity. It's tedious to keep telling NPCs to read more after a level-up so they should just keep reading until they can't gain any more skill. And it they have more books in their inventory they should switch to another book when they finish the first one.

**Comment by RenechCDDA** [comment]
>B) If you have enough tools and materials and you have N skilled helpers you should be able to craft N+1 recipes at the same time.

I agree but I should note that currently, sufficiently skilled NPCs assist you with tasks if they are skilled enough and in range. The first can drop you to 75% (~33% "increase" in speed), the second to 60%, and that's... it. https://github.com/CleverRaven/Cataclysm-DDA/blob/567594a0899ebfc7ed64df803cb26b952f7612bb/src/recipe.cpp#L126-L130

So what you are proposing in this regards would make NPCs much stronger for the player to acquire. Not saying that it shouldn't happen, just... consider the balance part of the equation. 

Personally I would say that something on the order of 85% efficiency per NPC (first crafts at base speed, second is 85%, third is 72.25%, etc...) would be appropriate. Although economy of scale is a very real thing, a *zone* does not require infrastructure - it's just a bunch of survivors shoved into a room, or sitting in an open field, etc with all the stepping on each others' toes that entails. Efficiencies of 100%+ for multiple NPCs should probably be limited to dedicated camps or other facilities.

>When an NPC starts crafting, reserve the tools that they need using item vars. I believe this is already done for other activities like construction, vehicle repair and vehicle disassembly.

Unfortunately it is *not* done for faction camps, which also would very much like to use such zones. Not that such an implementation needs to respect or nicely fit with faction camps' flow, but something to keep in mind.

**Comment by eltank** [comment]
Regarding NPC efficiency. I first though of this idea in the context of party-based gameplay (#51632). If I can freely switch between characters and they can all do the same thing it makes no sense to penalize some of them for doing stuff at the same time as another character. I'm willing to completely disable this feature outside of party-based gameplay if that is seen as breaking "balance".

Note that currently when they're helping you craft they're not actually using any tools. If a recipe requires sawing and you only have 1 saw, you still get the crafting bonus from NPC helpers (because they're just assisting, not actually crafting in parallel).

**Comment by eltank** [comment]
Let me put it another way:
With this feature you can ask 2 NPCs to work on different crafting recipes in two different rooms in your base. They can't even see each other. Would you not expect both of them to work as fast as you would (given equals stats and skills)? And if they happen to both be working on the same recipe, would you expect them to slow down? What would possibly cause that?

**Comment by RenechCDDA** [comment]
I did not realize this was related to the proposed party feature. 

Ultimately, I am neither the person which decides whether something is "balanced" nor do I think 100% crafting across infinite NPCs (party or not) is *necessarily* problematic. Please understand that I'm not posting because I think the concept is flawed or shouldn't be added - I'd love to see this!

I agree that two NPCs having access to space and separate resources on opposite sides of a wall slowing down because the other exists is a *ludicrous* idea; the problem comes in crafting a system around accounting for access to such space and resources without unintuitive or nonsensical behavior. Some level of abstraction is necessary. 

To extend my previous example a bit the scalar need not kick in immediately, but say only after several crafting NPCs in the reality bubble (e.g. 1.0>1.0>1.0>1.0>0.85>0.7225>etc). This would abstractly 'account; for the idea that there's sufficient space in any reality bubble for at least 4(or any arbitrary number of) people to work freely without specifically accounting for the needs of such a task.

But that's just an *example*, a crafting zone need not work like that nor in any way that remotely resembles it. # of NPCs scaling crafting might be completely ignorable from a balance perspective. I simply think that new additions are best compared to existing content, like NPC crafting assistance, to sound out potential concerns before they arise. 

Put another way, the difference between 1 player+2NPCs currently is ~66.7% more speed if all three were crafting at base speeds in a crafting zone. But 1 player + *20* NPCs is much more significant. Most players won't have or be willing to keep 20 NPCs around, in my opinion. But what about five? Or... seven? Is that average? Is the increase a concern at that point? These aren't questions that are intended to have direct answers, they're just useful *sounding boards* for consideration when dealing with new additions.

**Comment by PatrikLundell** [comment]
If this is implemented (which sounds like a good idea), I believe an option to just assemble the crafting in progress item and then immediately leave it at 0% complete beside the "normal" one would be useful.

When it comes to crafting efficiency, there are several aspects:
- Companions assisting with a reduced efficiency makes sense for single crafts as well as batches. This would mean I'd be able to complete this single needed item in one week rather than two, with the increased crafting times recently introduced. Sure, if the companions were skilled enough to do the job, you could farm it off to them, but since the PC tends to be the most skilled character the PC may be the only one having enough skills to actually lead the crafting effort. Thus, assisted crafting still has a use case.
- Each companion ought to craft at the full speed based on that companion's skills (assuming tool, etc. availability).
- Congestion in the workshop is a real world issue, but I don't know how you'd model it. It certainly doesn't make sense for people spread out at different locations within the bubble, and enforcing congestion in the workshop may lead to tedious micro management where you have to lead each companion away to "their" crafting station (which may be needed anyway, to get access to different fixed resources, such as forges and stoves, etc., unless companions get the smarts to seek these out on their own if handed a crafting object requiring such resources).
- The reality bubble is fairly large when it comes to how many people you can have within it without them stepping on each others' toes, so I think you'd run into an unmanageably large horde of companions well before running out of space for them to work on their own things without getting in each others' way (and I think it's useful to ask questions like RenechCDDA does). I had more than I really wanted to manage when I had the starting NPC->companion, the Refugee Center beggars, and a rescued prototype cyborg in 0.E2 stable.

**Comment by RenechCDDA** [comment]
@eltank This topic is also highly relevant to your proposal: #49472

**Comment by eltank** [comment]
> @eltank This topic is also highly relevant to your proposal: #49472

Thanks, I've added a couple of extra problems I saw discussed in there that I hadn't considered.
